helix-core/src/main/java/org/apache/helix/model/InstanceConfig.java

```diff
@@ -328,75 +462,192 @@ public String getInstanceDisabledReason() {
-    _record.setSimpleField(InstanceConfigProperty.INSTANCE_OPERATION.name(),
-        operation == null ? "" : operation.name());
-    if (operation == null || operation == InstanceConstants.InstanceOperation.ENABLE
-        || operation == InstanceConstants.InstanceOperation.DISABLE) {
+  public void setInstanceOperation(InstanceOperation operation) {
+    List<InstanceOperation> deserializedInstanceOperations = getInstanceOperations();
+
+    if (operation.getSource() == InstanceConstants.InstanceOperationSource.ADMIN) {
+      deserializedInstanceOperations.clear();
+    } else {
+      // Remove the instance operation with the same trigger if it exists.
+      deserializedInstanceOperations.removeIf(
+          instanceOperation -> instanceOperation.getSource() == operation.getSource());
+    }
+    if (operation.getOperation() != InstanceConstants.InstanceOperation.ENABLE) {
+      deserializedInstanceOperations.add(operation);
+    }
+    // Set the actual field in the ZnRecord
+    _record.setListField(InstanceConfigProperty.HELIX_INSTANCE_OPERATIONS.name(),
+        deserializedInstanceOperations.stream().map(instanceOperation -> {
+          try {
+            return _objectMapper.writeValueAsString(instanceOperation.getProperties());
+          } catch (JsonProcessingException e) {
+            throw new HelixException(
+                "Failed to serialize instance operation for instance: " + _record.getId()
+                    + " Can't set the instance operation to: " + operation.getOperation(), e);
+          }
+        }).collect(Collectors.toList()));
+
+    // TODO: Remove this when we are sure that all users are using the new InstanceOperation only and HELIX_ENABLED is removed.
+    if (operation.getOperation() == InstanceConstants.InstanceOperation.DISABLE) {
       // We are still setting the HELIX_ENABLED field for backwards compatibility.
       // It is possible that users will be using earlier version of HelixAdmin or helix-rest
       // is on older version.
-      // TODO: Remove this when we are sure that all users are using the new field INSTANCE_OPERATION.
-      setInstanceEnabledHelper(!(operation == InstanceConstants.InstanceOperation.DISABLE));
+
+      if (_record.getBooleanField(InstanceConfigProperty.HELIX_ENABLED.name(), true)) {
+        // Check if it is already disabled, if yes, then we don't need to set HELIX_ENABLED and HELIX_ENABLED_TIMESTAMP
+        setInstanceEnabledHelper(false, operation.getTimestamp());
+      }
+
+      _record.setSimpleField(InstanceConfigProperty.HELIX_DISABLED_REASON.name(),
+          operation.getReason());
+      _record.setSimpleField(InstanceConfigProperty.HELIX_DISABLED_TYPE.name(),
+          InstanceConstants.InstanceOperationSource.instanceOperationSourceToInstanceDisabledType(
+              operation.getSource()).name());
+    } else if (operation.getOperation() == InstanceConstants.InstanceOperation.ENABLE) {
```

Comment:
We need to think about the backward compatible behavior.

Say the instance was disabled in old field, but no new instance operations. Then we should let it be disabled.


My suggestion for this implementation to get old field disable time and merge in the new operation stack. Then we can populate it right and not have these branches.

Reply:
Should we write back to old fields when setting InstanceOperations with new API? That is the only way that a user of an old version of HelixAdmin can now if the instance is disabled from the new API.